### ISSUE TITLE
Python Linting

### ISSUE THREAD

**Issue by satiowadahc** [LinuxCNC/linuxcnc#1058]
Tracking Pylint process here on Master Branch.

Running 
```shell
  PL_FILES=$(find -type f -name "*.py" ! -path "**/.venv/**")
  pylint --unsafe-load-any-extension=y --output-format=colorized $PL_FILES
```
src/ Rated at -5.65/10
[pylint-report.txt](https://github.com/LinuxCNC/linuxcnc/files/5874770/pylint-report.txt)

lib/ Rated at -1.08/10
[pylint-report-lib.txt](https://github.com/LinuxCNC/linuxcnc/files/5874789/pylint-report-lib.txt)

Admittedly at lot of these will be related to lack of doc strings, white space and python2 syntaxes. But on the whole its another metric to evaluate code bases, remove logical errors and increase efficiency. 

Thoughts, Concerns, Comments?

**Comment by satiowadahc** [comment]
Looking at buildbot, could be a check, once the move to python 3 is complete. 

http://docs.buildbot.net/latest/manual/configuration/steps/pylint.html

I usually strive for at least 8/10 on PyLint, which could take some work to get to, but would be nicer in the long run.

**Comment by c-morley** [comment]
What is the risk/benefit of pylint (I've never used it - just googled it)
One article quote: 'Why Pylint is both useful and unusable'  seemed pertinent

Linuxcnc has no hard standard for coding - in general it's preferred to follow the original code style.
I don't think we have the luxury of turning away code because of style.

**Comment by satiowadahc** [comment]
The big things pylint has saved me on is name shadowing, and variable contexts. You can customize it to ignore certain errors such as formatting, modules not being part of a package, non-subscriptable variables. In raw format like I ran it is definitely not useful. But then someone or a group has to decide which things the community doesn't want to inspect, and what they do want to inspect.

EG in qt_action there is a nfile variable that doesn't get assigned in some contexts and I seen it on pylint, but only became a problem when I tried to save a file with 2 periods in the name. (Been meaning to put a pr for the fix, but ended up just fixing it on my GUI)

TL;DR
Benefits - Code Quality, Error Catching,
Risks - One more thing to maintain, initial configuration would need to be decided on

**Comment by c-morley** [comment]
another risk is offered code not accepted and the author gives up trying to get it in. (A problem we have even without code standards)

And we'd have to fight about code standards :) (I find PEP8 a little too picky about some things)

I personally believe we need more active developers before taking things like this on.

Oh and I'd be interested in the found error in qt_action ....

**Comment by satiowadahc** [comment]
I mean we don't necessarily need to reject code on it. If we can get the code to the positives, humans in general usually strive for higher ratings. It just gives incentives to increase the number. And each addition would hardly affect it, I think LCNC is around 1.3 million lines of code iirc, so each little line hardly affects the actual rating 

 True. Would it be opposed if I keep going through and trying to get it the rating a little higher? I mean as selfish as it is, I tend to only touch the part of the code base I end up pulling for my mills and interfaces. 

Once I get to work this morning, I can get a PR up for it. (PR 983, would you have time to look at again? (:  )

**Comment by c-morley** [comment]
Either the developer fixes it when it's offered, asks the author to fix it before it's accepted or someone like yourself comes along and fixes a bunch of things all at once.
The first one takes man power we don't have.
Seconds relies on the author willing and capable of fixing.
The third causes big merge problems (when a whole bunch of files are done at once).
I don't know what the best answer is.

To be clear we want bug fixes and clean code is better - it usually is better to fix code style only when adding/fixing something else.
If the styling changes are large it should be separate if very small then it can be included.

We have gone this far - If you have more code changes let's do it - we rarely have someone that interested :)
I had done a PEP8 audit and fix on QtVCP but that was before the code officially was included in linuxcnc so it didn't annoy too many people.

PR 983 - yes sorry for delay

**Comment by satiowadahc** [comment]
I'm coming back to this as I have to review my own CI testing again, and finally have my own computer set up again. If I were to start this process again would smaller commits be preferred or one file at a time? 

One thing for sure I would like to remove is shadow names. EG. Format is one that comes to mind that's being used as a variable but it's also a string function.

I propose we add the check but don't enforce it. Developers tend to be people who like to optimize things. If I see a metric I want to make it a better score and its slightly insulting if I lower it.

**Comment by c-morley** [comment]
Adding the check for shadow names seems a good test to add - I'm game.

**Comment by ClariNerd617** [comment]
The solution is clearly to agree to a single style that can be imposed with a formatter such as black or ruff, and then add that to the [pre-commit hooks](https://pre-commit.com/), as that way it becomes an automated solution as well as self-enforcing. It also forces checking to make sure new patches don't cause regressions, another plus.
